src/plinder/data/splits.py

```diff
@@ -644,88 +652,240 @@ def assign_split_membership(
-            to_add.extend(clusters[c][: cfg.split.num_val_representatives])
-            if len(clusters[c]) > cfg.split.num_val_representatives:
-                to_remove.extend(clusters[c][cfg.split.num_val_representatives :])
-        val_system_ids.extend(to_add)
-        remove_ids.extend(to_remove)
-        start_cluster_index += 1
-    # train_clusters = list(
-    #     set(
-    #         entries[entries["split"] == "train"]["cluster_for_val_split"].sample(
-    #             frac=1.0
-    #         )
-    #     )
-    # )
-    # val_clusters = set(train_clusters[: len(train_clusters) // 10])
-    # train_system_ids = set(entries[entries["split"] == "train"]["system_id"])
-    # val_system_ids = set(
-    #     [x for x in train_system_ids if system_to_cluster_val[x] in val_clusters]
-    # )
+    remove_ids -= set(val_system_ids)
+    LOG.info(
+        f"val_system_ids after max_val of {cfg.split.num_val}: {len(val_system_ids)}"
+    )
+
+    # Remove val PDB IDs
+    for x in val_system_ids:
+        remove_ids |= pdb_id_to_systems[x.split("__")[0]]
+    remove_ids -= set(val_system_ids)
+
     entries.loc[entries["system_id"].isin(set(val_system_ids)), "split"] = "val"
-    LOG.info(f"Additional {len(set(remove_ids))} to be removed due to val")
-    entries.loc[entries["system_id"].isin(set(remove_ids)), "split"] = "removed"
+    LOG.info(f"Additional {len(remove_ids)} to be removed due to val")
+    entries.loc[entries["system_id"].isin(remove_ids), "split"] = "removed"
+
+    # Add back systems in removed with novel ligands compared to train
+    train_val_ligand_clusters = set(
+        entries[entries["split"].isin(["train", "val"])][
+            f"{cfg.split.ligand_cluster_metric}__{cfg.split.ligand_cluster_threshold}__{cfg.split.ligand_cluster_cluster}"
+        ]
+    )
+    removed_ligand_clusters = set(
+        entries[
+            (entries["split"] == "removed")
+            & (entries["system_pass_validation_criteria"])
+            & (entries["system_pass_statistics_criteria"])
+        ][
+            f"{cfg.split.ligand_cluster_metric}__{cfg.split.ligand_cluster_threshold}__{cfg.split.ligand_cluster_cluster}"
+        ]
+    )
+    novel_ligand_clusters = removed_ligand_clusters - train_val_ligand_clusters
+    test_ccd_codes = set(
+        entries[entries["split"] == "test"]["system_proper_ligand_unique_ccd_codes"]
+    )
+    novel_ligand_systems = set(
+        entries[
+            entries[
+                f"{cfg.split.ligand_cluster_metric}__{cfg.split.ligand_cluster_threshold}__{cfg.split.ligand_cluster_cluster}"
+            ].isin(novel_ligand_clusters)
+            & (~entries["system_proper_ligand_unique_ccd_codes"].isin(test_ccd_codes))
+            & (entries["split"] == "removed")
+            & (entries["system_pass_validation_criteria"])
+            & (entries["system_pass_statistics_criteria"])
+        ]
+        .groupby(["entry_pdb_id", "system_proper_ligand_unique_ccd_codes"])
```

Comment:
should we group here only by `"system_proper_ligand_unique_ccd_codes"` to only be adding a novel ligand once?
Here it can add as many as there are but from different PDB?

Reply:
Captured this comment in a TODO for future enhancements